### DISCUSSION TITLE
Machine that was originally created without device aliases (pre 0.6.0)

### DISCUSSION THREAD

**Discussion by Jeansen** [vagrant-libvirt/vagrant-libvirt#1552] [Q&A] [answered]
I made a big jump from 0.4.1 to 0.10.1. Now, I am seeing this in my Vagrant output:

```
==> bios-lvm-all-in-one-test: Running trigger...
==> bios-lvm-all-in-one-test: Adding synced folders.
==> bios-lvm-all-in-one-test: Machine that was originally created without device aliases (pre 0.6.0), using fall-back approach for device identification.
==> bios-lvm-all-in-one-test: 
Traceback (most recent call last):
	29: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/batch_action.rb:86:in `block (2 levels) in run'
	28: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/machine.rb:203:in `action'
	27: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/machine.rb:203:in `call'
	26: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/environment.rb:631:in `lock'
	25: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/machine.rb:217:in `block in action'
	24: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/machine.rb:248:in `action_raw'
	23: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/runner.rb:101:in `run'
	22: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/util/busy.rb:19:in `busy'
	21: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/runner.rb:101:in `block in run'
	20: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builder.rb:180:in `call'
	19: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	18: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builtin/trigger.rb:32:in `call'
	17: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	16: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builtin/config_validate.rb:25:in `call'
	15: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	14: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builtin/box_check_outdated.rb:36:in `call'
	13: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	12: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builtin/call.rb:53:in `call'
	11: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/runner.rb:101:in `run'
	10: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/util/busy.rb:19:in `busy'
	 9: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/runner.rb:101:in `block in run'
	 8: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builder.rb:180:in `call'
	 7: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	 6: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:127:in `block in finalize_action'
	 5: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	 4: from /media/marcel/5f7ca5f7-a2e6-41ea-86af-a1c4c7b0aa13/bcrm_test/.vagrant.d/gems/2.7.6/gems/vagrant-libvirt-0.10.1/lib/vagrant-libvirt/action/cleanup_on_failure.rb:21:in `call'
	 3: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
	 2: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/builtin/provision.rb:80:in `call'
	 1: from /opt/vagrant/embedded/gems/2.3.0/gems/vagrant-2.3.0/lib/vagrant/action/warden.rb:48:in `call'
/media/marcel/5f7ca5f7-a2e6-41ea-86af-a1c4c7b0aa13/bcrm_test/.vagrant.d/gems/2.7.6/gems/vagrant-libvirt-0.10.1/lib/vagrant-libvirt/action/resolve_disk_settings.rb:103:in `call': undefined method `[]' for nil:NilClass (NoMethodError)
```

The machine still works, but I wonder how I could resolve this warning (error). Any help is highly appreciated.

**Comment by electrofelix** [comment]
Would you be able to generate a debug log for this? The fallback code for handling older machines is still in place so it should work. I can only assume I've missed a test scenario when I was considering the different states that could exist.

**Comment by electrofelix** [comment]
My guess is that the following block of code is not doing what I expected
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/591e151b13f9b520618854a245af2489de677352/lib/vagrant-libvirt/action/resolve_disk_settings.rb#L67-L73

To solve, I expect I'll need to see the domain XML definition.

**Comment by Jeansen** [comment]
@electrofelix I\ll attach the log soon. The Vagrant script is quite generic, but then you'll have the full run and can have a look.

**Reply by electrofelix** [reply]
Cheers, the log should have the domain xml included in it as well, so no need to worry about that then

**Comment by Jeansen** [comment]
yes, I can see it. Script will take some time, still. First it creates some machines, clones them and then there is a test phase. Exactly then, this error occurs. But you'll see, when I'll upload it.

**Comment by Jeansen** [comment]
Sorry, I got a bit distracted. Here's the gist to the log file: https://gist.github.com/Jeansen/91660afa0ead5c421fade0c31b621baa

**Comment by electrofelix** [comment, marked as answer]
Ah, I see what is happening, it's not that the machine was brought up with an old version of the plugin, but rather the more recent code has an assumption that the root domain disk will have an aliases associated with it, while the script that was run is detaching a disk and reattaching a replacement without any alias.

```
"/usr/bin/bash", "-c", "export LIBVIRT_DEFAULT_URI=qemu:///system;\n              virt-xml bcrm_test_bios-lvm-all-in-one-test --edit model=lsilogic --controller model=virtio-scsi;\n              virsh detach-disk bcrm_test_bios-lvm-all-in-one-test sda --config;\n              virsh attach-disk bcrm_test_bios-lvm-all-in-one-test /media/marcel/5f7ca5f7-a2e6-41ea-86af-a1c4c7b0aa13/bcrm_test/disks/libvirt/bios-lvm-all-in-one/disk2_1-1.qcow2 sda --targetbus scsi --driver qemu --subdriver qcow2 --type disk --config"
```

Additionally the fallback code in turn is expecting that the default domain volume coming from a box should use the `.img` suffix based on what vagrant-libvirt would upload into libvirt. This is to distinguish from additional storage devices being attached via the Vagrantfile that are not part of the box domain and should in certain cases not be removed. But to identify the correct ones, also need to be able to identify the domain volume(s) that are needed by the machine to boot and function.

The quick fix is to add `--alias ua-box-volume-0` to the attach-disk call.

I'm trying to think if there is anyway to safely handle this scenario automatically, and not sure whether there are any good answers. One option might be to check the number of disks in the domain XML and if there are no additional storage devices defined, just assume they are all domain volumes in the event there is no alias found. Though that may go wrong in the case that someone has removed an additional storage entry from the Vagrantfile.

**Reply by Jeansen** [reply]
OK, that works. But using a different alias gave me a traceback, though it seems to work, too.

**Reply by electrofelix** [reply]
It needs to be that specific alias. When vagrant-libvirt creates a machine from a box definition, in order to support boxes having more than one disk image, and still being able to safely destroy them without accidentally deleting any of the additional storage configured to be persistent, it assigns each disk image from the box an alias in the libvirt domain. Starting with ua-box-volume-0 and incrementing the final digit for each disk as needed. For the additional storage, they start with the alias ua-disk-volume-0, similarly incrementing the last digit as needed.

So unfortunately the alias is specific. Basically need to include the same alias as was given to the disk image that was detached before attaching your custom disk image.

**Comment by Jeansen** [comment]
Oh, great. I'll check that. I konw, my script is quite ... special. Evolved over time, but this way I am able to run different scenarios when cloning systems back and forth.

**Comment by Jeansen** [comment]
If I can support you with any way to make vagrant-libvirt more robust, please let me know.
